papers/numba/gufunc/gufunc.rst

```diff
@@ -0,0 +1,123 @@
+elements in each row of a 2D array, producing a 1D array as output.  A pure
+NumPy approach to this problem would look like:
+
+.. code-block:: python
+
+    def np_weighted_avg(x, weights):
+    return (x * weights).sum(axis=1)
+
+We assume that ``weights`` is a 1D array and has already been normalized so
+that ``weights.sum() == 1``. For a 10000x100 input array (producing a 10000
+element output array), the above function takes 3.6 ms on our test system.
+
+If we thinking about how NumPy works in the Python interpreter, we realize
+that the above implementation creates a temporary array with the weighted
+values of ``x``, only to immediately throw it away after summing along axis 1.
+Numba allows us to express the operation without temporaries:
+
+.. code-block:: python
+
+    @numba.jit(nopython=True)
+    def nb_weighted_avg_loops(x, weights):
+        result = numpy.empty(x.shape[0], x.dtype)
+        for i in range(x.shape[0]):
+
+            row_sum = 0.0
+            for v, w in zip(x[i], weights):
+                row_sum += v*w
+                
+            result[i] = row_sum
+            
+        return result
+
+This is much how one would write this function in C or FORTRAN, although we
+can use some Python conviences functions like ``zip`` to more naturally loop
+through each row and the ``weights`` array at the same time.  This version of
+the function runs in 1.5 ms, or an improvement of 2.6x over NumPy.
+
+Now, we might be content to stop here, but this form of the function has some
+disadvantages.  First, the explicit looping limits this implementation to only
+work on 2D arrays.  If we had a 3D array (say 1000x1000x100), we would have to
+write another version of the above function to operate on 3D arrays.  Second,
+it would be nice to make this more like a ufunc so we do not have to manage
+allocating the output array, and so we can focus on core piece of the
+algorithm: the weighted sum itself.
+
+NumPy has a generalization of the ufunc, called a *gufunc*, that Numba can
+also create.  A gufunc is distinguished from a ufunc by having a layout
+signature that describes the dimensionality of the inputs that should be
+passed to the core function.  The signature is a mini-language described more
+fully in the NumPy documentation, though we will give some examples below.  A
+regular ufuncs with two input arguments implicitly have a signature of
+``(),()->()``, indicating that the core function takes two scalar inputs and
+produces a scalar output, for example.
+
+Numba creates gufuncs using the ``@guvectorize`` decorator.  Unlike the
+``@vectorize`` decorator for making ufuncs, ``@guvectorize`` always requires a
+type signature (along with the gufunc signature) for the function inputs.  If
+we wish to compile our weighted average gufunc for float32 and float64 inputs,
+we would write the following (explained further below):
+
+.. code-block:: python
+
+    @numba.guvectorize(['(float64[:], float64[:], float64[:])', 
```

Comment:
IIRC, signatures are tried in order by Numpy, so putting the float64 signature before the float32 risks always converting SP arrays to DP, right?